UserBalanceMicroservice/UserBalanceMicroservice/UserBalanceMicroservice/BalanceContext.cs

```diff
@@ -0,0 +1,109 @@
+﻿using eventList.CustomTypes;
+using Microsoft.Extensions.Options;
+using MongoDB.Driver;
+using UserBalanceMicroservice.Configs;
+using UserBalanceMicroservice.Models;
+namespace UserBalanceMicroservice
+{
+    public class BalanceContext
+    {
+        private readonly IMongoDatabase _mongoDatabase;
+
+        private const string _userCollection = "User";
+        private const string _transactCollection = "Transact";
+        public BalanceContext(
+            IOptions<DatabaseSettings> bookStoreDatabaseSettings)
+        {
+            var mongoClient = new MongoClient(
+                bookStoreDatabaseSettings.Value.ConnectionString);
+
+            _mongoDatabase = mongoClient.GetDatabase(
+                bookStoreDatabaseSettings.Value.DatabaseName);
+
+        }
+
+        public async Task<bool> ChangeBalanceAsync(string idUser, decimal sum, bool isAdd, string transactionId)
+        {
+            var user = await GetOrCreateUserAsync(idUser);
+            string lastTransactId = user.LastTransactId;
+            var tempMoney = user.Money;
+
+            if(user.LastTransactId == transactionId) { return false; }
+
+            var filter = Builders<User>.Filter.Where(x => x.Id == idUser
+                    && x.LastTransactId == lastTransactId
+                    && x.Money == tempMoney);
+
+            if (isAdd)
+            {
+                user.Money += sum;
+            }
+            else
+            {
+                if (user.Money < sum)
+                {
+                    return false;
+                }
+                user.Money -= sum;
+            }
+
+            var update = Builders<User>.Update.Set(x => x.Money, user.Money);
+            update.AddToSet("LastTransactId", transactionId);
+
+            var temp = await _mongoDatabase.GetCollection<User>(_userCollection)
+                    .FindOneAndUpdateAsync(filter, update);
+            if (temp != null)
+            {
+                await SetTransactAsync(idUser, isAdd, transactionId);
+                return true;
+            }
+            return false;
```

Comment:
```suggestion
            if (temp == null)
                return false
            await SetTransactAsync(idUser, isAdd, transactionId);
            return true;
```
Такое проще читается